### ISSUE TITLE
Non-deterministic behavior related to multiprocessing

### ISSUE THREAD

**Issue by danail-k** [Microsoft/LightGBM#2085]
On certain train datasets and certain values of n_jobs training becomes non-deterministic. Setting n_jobs to specific values (1, 2, 8) produces the same results across different machines. However, setting it to most other values returns different models even on the same machine and in the same python process. I set subsampling to 1, as it is known to cause a similar issue and I pass a seed to the classifier. This happened on multiple ubuntu machines with different hardware and with lightgbm versions 2.1.2, 2.2.2 and 2.2.3, using python 3.6. The problem seems to be only in training - afterwards every model predicts deterministically on different machines.

So far the smallest dataset I can reproduce it with is of 200 rows and 1 col.

```import numpy as np
from lightgbm import LGBMClassifier

clf = LGBMClassifier(
    num_leaves=93,
    learning_rate=0.04,
    n_estimators=500,
    class_weight='balanced',
    min_child_samples=20,
    subsample=1.0,
    subsample_freq=0,
    colsample_bytree=1.0,
    random_state=1337,
    n_jobs=10,
    max_bin=127,
    verbose=-1)

X = np.array([[ 5.96745376e-05], [ 0.00000000e+00], [ 3.26438784e-03], [ 0.00000000e+00], [ 0.00000000e+00], [-2.00604834e-03], [ 6.89087529e-03], [ 6.89087529e-03], [ 7.45642537e-05], [ 0.00000000e+00], [-2.30295642e-04], [ 2.22012866e-03], [ 0.00000000e+00], [-4.60559415e-04], [ 0.00000000e+00], [ 1.04821650e-02], [ 0.00000000e+00], [ 1.10302710e-04], [ 4.25727194e-04], [ 4.25727194e-04], [-3.19723797e-04], [-3.19723797e-04], [ 0.00000000e+00], [ 0.00000000e+00], [ 5.96745376e-05], [ 0.00000000e+00], [ 0.00000000e+00], [ 6.68999134e-03], [ 0.00000000e+00], [-8.40502544e-05], [-9.35432443e-04], [ 1.97680132e-03], [-5.43681066e-03], [ 2.80661479e-04], [-3.19723797e-04], [ 2.27653494e-04], [ 7.80984759e-03], [ 2.27653494e-04], [ 6.25655695e-04], [-4.42358525e-03], [ 1.04821650e-02], [ 0.00000000e+00], [ 0.00000000e+00], [-4.42358525e-03], [ 0.00000000e+00], [ 6.25655695e-04], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [-4.60559415e-04], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [-4.60559415e-04], [ 0.00000000e+00], [ 2.27653494e-04], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 1.00493453e-04], [ 5.36797452e-04], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [-4.60559415e-04], [ 0.00000000e+00], [ 0.00000000e+00], [ 5.14005171e-03], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [-4.60559415e-04], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [-4.60559415e-04], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [-4.60559415e-04], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [-4.60559415e-04], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 9.03615146e-05], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 2.27653494e-04], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 0.00000000e+00], [ 4.44649777e-06], [ 0.00000000e+00], [ 0.00000000e+00], [-4.60559415e-04], [ 0.00000000e+00], [ 0.00000000e+00], [-1.80223200e-04], [-7.73051943e-05], [ 0.00000000e+00], [ 0.00000000e+00], [ 1.10302710e-04], [-1.14844227e-03], [ 0.00000000e+00], [ 5.96745376e-05], [ 0.00000000e+00], [ 3.26438784e-03], [ 0.00000000e+00], [-2.00604834e-03], [ 6.89087529e-03], [ 6.89087529e-03], [ 7.45642537e-05], [ 0.00000000e+00]])
y = [4, 4, 4, 4, 4, 4, 5, 4, 8, 4, 4, 8, 4, 4, 4, 4, 4, 6, 0, 7, 0, 7, 4, 4, 4, 4, 4, 4, 3, 4, 4, 4, 4, 4, 4, 4, 1, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 2, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4]

clf.fit(X, y)
predictions = clf.predict_proba(X)```

# ^ predictions end up being different.

**Comment by guolinke** [comment]
@danail-k 
I guess the root cause is the `reduce_sum` by openmp. As many metric calculations contain sum-up of error/accuracy of each data point, the `reduce_sum` is used for speed-up.
when n_jobs is the power of 2, the reduce_sum produce the same results.

To verify this, you can try the `auc` metric, which should not be affected by reduce_sum.

**Comment by danail-k** [comment]
@guolinke 
Thanks for the quick response. I'm not sure I understand how the evaluation metric changes the predicted probabilities. Also I get non-deterministic results with other powers of 2, like 16 and 32 and setting metric='auc' is incompatible with multiclass objective.

**Comment by guolinke** [comment]
@danail-k 
Sorry, I misunderstood your problem. 
I will investigate this issue.

**Comment by guolinke** [comment]
@danail-k 
Finally, I found the root cause, refer to the code: 
https://github.com/Microsoft/LightGBM/blob/9de526f8acc36f493252c635880bca043e643ae3/src/treelearner/leaf_splits.hpp#L63-L67

The result is deterministic after removing the omp here.

there are two problems: 
1. the imbalance class weights:
```
weight = [ 0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 , 22.22222222,  0.1175779 , 11.11111111,  0.1175779 ,
        0.1175779 , 11.11111111,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 , 22.22222222, 11.11111111, 11.11111111,
       11.11111111, 11.11111111,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 , 22.22222222,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 , 22.22222222,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 , 22.22222222,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,
        0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ,  0.1175779 ]
```

2. the multi-threading sum-up of "gradient * weight", whose value ranges are very different.

The multi-threading floating sum-up cannot ensure the deterministic for the data in very different ranges. 

BTW, although it is not deterministic, the difference is `4e-16`.
